proposals/p6357.md

```diff
@@ -0,0 +1,400 @@
+    -   C++ types are mapped to Carbon types with the same representation. This
+        avoids the need for any deep copies on the interop boundary.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   The convenience names `Cpp.void` and `Cpp.nullptr` make it easier to
+        write code that interoperates with C++ void pointers and C++ nullable
+        pointers.
+-   [Practical safety and testing mechanisms](/docs/project/goals.md#practical-safety-and-testing-mechanisms)
+    -   C++ null pointers cannot "leak" into Carbon non-nullable pointer types,
+        except by violating a "nonnull" annotation in the C++ code. If
+        `_Nonnull` is violated, the result is not undefined behavior unless the
+        pointer is used in a context that would result in undefined behavior in
+        C++, such as a load or store through the pointer.
+-   [Modern OS platforms, hardware architectures, and environments](/docs/project/goals.md#modern-os-platforms-hardware-architectures-and-environments)
+    -   We do not assume that a null pointer has an all-zero-bits
+        representation. This is not true in practice on some modern
+        heterogeneous compute architectures.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   This proposal significantly enhances interoperability with C++ code, by
+        providing a mapping for very common vocabulary types in C++.
+
+Principles:
+
+-   [All APIs are library APIs](/docs/project/principles/library_apis_only.md)
+    -   The types added to support C++ are all ordinary class types provided in
+        the Carbon prelude.
+-   [Low context-sensitivity](/docs/project/principles/low_context_sensitivity.md)
+    -   The Carbon `Cpp.void` type has only one meaning, rather than having a
+        context-sensitive meaning. The C++ `void` type still has two different
+        meanings, and therefore two different mappings into Carbon, but that
+        problem is outside our domain.
+-   [Namespace cleanliness](/docs/project/principles/namespace_cleanliness.md)
+    -   The names we are adding to the `Cpp` package, `void` and `nullptr`, are
+        C++ keywords, and so do not conflict with any C++ identifier. While
+        Clang does provide an extension to define entities with these names, for
+        example `int __identifier(void);`, interoperting with such code is not a
+        priority.
+
+## Future work
+
+-   Add a design for `Core.Optional`, as we now have multiple language proposals
+    that depend upon it.
+-   Design interop for function pointers.
+-   Design interop for pointers to members.
+-   Design an approach for providing the functionality that C++ exposes as
+    pointer indexing and pointer arithmetic.
+
+## Alternatives considered
```

Comment:
One alternative that might be good to consider is the idea of mapping `void*` to something like `u8*` or some `byte*` type.

That would also be a good place to surface why it ends up important for these automatically mapped types to be 1:1 instead of N:1 (like both `void*` and `std::byte*` mapping to a Carbon `byte*`), where we would lose type information when round tripping from a Carbon generic back into a C++ template, and might not even have the needed specializations in C++ available.

Reply:
Added a section